File: app/android/app/src/main/java/com/harvestschools/app/HomeWidgetPlugin.java
package com.harvestschools.app;

import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;

@CapacitorPlugin(name = "HomeWidget")
public class HomeWidgetPlugin extends Plugin {

    @PluginMethod
    public void isSupported(PluginCall call) {
        JSObject result = new JSObject();
        result.put("value", true);
        call.resolve(result);
    }

    @PluginMethod
    public void setQuickActions(PluginCall call) {
        String payload = call.getString("payload");

        if (payload == null || payload.isEmpty()) {
            call.reject("A payload is required");
        } else {
            HarvestWidgetStore.saveQuickActions(getContext(), payload);
            QuickActionsWidgetProvider.refreshAllWidgets(getContext());
            JSObject result = new JSObject();
            result.put("value", true);
            call.resolve(result);
        }
    }
}

File: app/ios/App/App/HomeWidgetPlugin.swift
import Foundation
import Capacitor
import WidgetKit

@objc(HomeWidgetPlugin)
public class HomeWidgetPlugin: CAPPlugin, CAPBridgedPlugin {

    public let identifier = "HomeWidgetPlugin"

    public let jsName = "HomeWidget"

    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "isSupported", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "setQuickActions", returnType: CAPPluginReturnPromise)
    ]

    @objc func isSupported(_ call: CAPPluginCall) {
        call.resolve(["value": UserDefaults(suiteName: harvestAppGroupIdentifier) != nil])
    }

    @objc func setQuickActions(_ call: CAPPluginCall) {
        guard let payload = call.getString("payload") else {
            call.reject("A payload is required")

            return
        }

        guard let sharedDefaults = UserDefaults(suiteName: harvestAppGroupIdentifier) else {
            call.reject("The shared app group is not available")

            return
        }

        sharedDefaults.set(payload, forKey: harvestQuickActionsKey)

        WidgetCenter.shared.reloadAllTimelines()

        call.resolve(["value": true])
    }
}
